src/linter.cc

```diff
@@ -77,39 +80,136 @@ absl::Status checkLineLength(absl::string_view sql, int lineLimit,
-    // TODO(orhan_uysal): Implement a function checking
-    // if all characters in ASTNode(keyword node) is uppercase
+absl::Status checkSemicolon(absl::string_view sql) {
+    absl::Status return_status = absl::OkStatus();
+    std::unique_ptr<zetasql::ParserOutput> output;
+
+    zetasql::ParseResumeLocation location =
+        zetasql::ParseResumeLocation::FromStringView(sql);
+    bool isTheEnd = false;
+    while ( !isTheEnd && return_status.ok() ) {
+        return_status = zetasql::ParseNextScriptStatement(
+            &location, zetasql::ParserOptions(), &output, &isTheEnd);
+        if ( !isTheEnd && return_status.ok() ) {
+            int endLocation = output -> statement() -> GetParseLocationRange()
+                .end().GetByteOffset();
+
+            if ( endLocation < sql.size() && sql[endLocation] != ';' ) {
+                return absl::Status(
+                        absl::StatusCode::kFailedPrecondition,
+                        absl::StrCat(
+                            "Each statemnt should end with a consequtive",
+                            "semicolon ';'"));
+            }
+        }
+    }
+    return return_status;
+}
+
+bool allUpperCase(const absl::string_view &sql,
+    const zetasql::ParseLocationRange &range) {
+    for (int i = range.start().GetByteOffset(); i < range.end().GetByteOffset(); i++) {
+        if ( 'a' <= sql[i] && sql[i] <= 'z' )
+            return false;
+    }
     return true;
 }
 
 absl::Status checkUppercaseKeywords(absl::string_view sql) {
-    std::vector<const zetasql::ASTNode *> *keywords;
-    *keywords = std::vector<const zetasql::ASTNode *>();
-
-    std::unique_ptr<zetasql::ParserOutput> output;
+    ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
+    std::vector<zetasql::ParseToken> parse_tokens;
+    absl::Status tokenizer_status =
+        GetParseTokens(ParseTokenOptions(), &location, &parse_tokens);
+
+    // Keyword definition in tokenizer is very wide,
+    // it include some special characters like ';', '*', etc.
+    // Keyword Uppercase check will simply ignore characters
+    // outside of english lowercase letters.
+    for ( auto &token : parse_tokens ) {
+        if ( token.kind() == zetasql::ParseToken::KEYWORD ) {
+            if (!allUpperCase(sql, token.GetLocationRange())) {
+                return absl::Status(
+                    absl::StatusCode::kFailedPrecondition,
+                    absl::StrCat("All keywords should be Uppercase, In character ",
+                    std::to_string(token.GetLocationRange().start().GetByteOffset()),
+                    " string should be: ", token.GetSQL()));
+            }
+        }
+    }
 
-    absl::Status parser_status = zetasql::ParseStatement(sql,
-            zetasql::ParserOptions(), &output);
+    return absl::OkStatus();
+}
 
-    if ( !parser_status.ok() )
-        return parser_status;
+absl::Status checkCommentType(absl::string_view sql) {
```

Comment:
I tried to make it with parsed tokens, but unfortunately (like you said) ZetaSQL parse // as 2 seperate Keyword tokens and try to tokenize everything comes afterwards. So, I will add checks for muiltiline comments and multiline string literals. (I hope we are not missing more cases)

Reply:
bah, github uses markdown by default o_O.. my previous message with comment types was supposed to be:

```
-- Comment
# Comment 
// Comment
/* Comment */
/*
Multi-line
comment
*/
```

You can skip for now dealing with `// Comment` style. Later, we can reuse code from the formatter (or reimplement own version): in formatter, we replace all '//' in the input with '--' before tokenising. Then restore '//' back in the final output.

Reply:
I implemented a function covering these types of comments, plus string literals that you said. I also thought about replacing '//' then tokenizing but I find this way easier because of situations like '/* comment1 */// comment 2'